src/vs/workbench/contrib/terminal/browser/terminalConfigurationService.ts

```diff
@@ -27,14 +32,18 @@ export class TerminalConfigurationService extends Disposable implements ITermina
 
 	constructor(
 		@IConfigurationService private readonly _configurationService: IConfigurationService,
+		@IContextKeyService private readonly _contextKeyService: IContextKeyService,
 	) {
 		super();
+		this._terminalSettingShellIntegrationEnabled = TerminalContextKeys.terminalSettingShellIntegrationEnabled.bindTo(this._contextKeyService);
+		this._terminalSettingShellIntegrationEnabled.set(this._configurationService.getValue(TerminalSettingId.ShellIntegrationEnabled));
```

Comment:
at startThe setting doesn't necessarily determine whether shell integration is enabled or not, just that SI injection is enabled which may or may not work depending on the terminal. The best context here would be if the terminal supports full cwd detection, ie. `CwdDetectionCapability` exists. I think best would be to have `CwdDetectionCapability.updateCwd/cwd` be the source of truth for the value, if that capability does not exist then set it to undefined when an instance is focused, and add a listener for when the capability activates.